loader/loader.go

```diff
@@ -104,32 +104,49 @@ func NewWorker(loader *Loader, id int) (worker *Worker, err error) {
 
 // Close closes worker
 func (w *Worker) Close() {
+	// simulate the case that doesn't wait all doJob goroutine exit
+	failpoint.Inject("workerCantClose", func(_ failpoint.Value) {
+		w.tctx.L().Info("", zap.String("failpoint", "workerCantClose"))
+		failpoint.Return()
+	})
+
 	if !atomic.CompareAndSwapInt64(&w.closed, 0, 1) {
+		w.wg.Wait()
+		w.tctx.L().Info("already closed...")
 		return
 	}
 
+	w.tctx.L().Info("start to close...")
 	close(w.jobQueue)
 	w.wg.Wait()
+	w.tctx.L().Info("closed !!!")
 }
 
-func (w *Worker) run(ctx context.Context, fileJobQueue chan *fileJob, workerWg *sync.WaitGroup, runFatalChan chan *pb.ProcessError) {
+func (w *Worker) run(ctx context.Context, fileJobQueue chan *fileJob, runFatalChan chan *pb.ProcessError) {
 	atomic.StoreInt64(&w.closed, 0)
-	defer workerWg.Done()
 
 	newCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	defer func() {
+		cancel()
+		// make sure all doJob goroutines exit
+		w.Close()
+	}()
 
 	ctctx := w.tctx.WithContext(newCtx)
 
 	doJob := func() {
-		defer w.wg.Done()
 		for {
 			select {
 			case <-newCtx.Done():
-				w.tctx.L().Debug("execution goroutine exits")
+				w.tctx.L().Info("context canceled, execution goroutine exits")
 				return
 			case job, ok := <-w.jobQueue:
-				if !ok || job == nil {
+				if job == nil {
+					w.tctx.L().Info("jobs are finished, execution goroutine exits")
+					return
+				}
+				if !ok {
```

Comment:
if `jobQueue` is closed, `job` is nil, and `ok` is false,  so it is better to put this statement before L144